src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEntry.Unix.cs

```diff
@@ -34,52 +33,42 @@ namespace System.IO.Enumeration
             entry._fullPath = ReadOnlySpan<char>.Empty;
             entry._fileName = ReadOnlySpan<char>.Empty;
 
-            // IMPORTANT: Attribute logic must match the logic in FileStatus
+            entry._status.InvalidateCaches();
+
+            bool isDirectory = directoryEntry.InodeType == Interop.Sys.NodeType.DT_DIR;
+            bool isSymlink   = directoryEntry.InodeType == Interop.Sys.NodeType.DT_LNK;
+            bool isUnknown   = directoryEntry.InodeType == Interop.Sys.NodeType.DT_UNKNOWN;
 
-            bool isDirectory = false;
-            bool isSymlink = false;
-            if (directoryEntry.InodeType == Interop.Sys.NodeType.DT_DIR)
-            {
-                // We know it's a directory.
-                isDirectory = true;
-            }
             // Some operating systems don't have the inode type in the dirent structure,
             // so we use DT_UNKNOWN as a sentinel value. As such, check if the dirent is a
-            // directory.
-            else if ((directoryEntry.InodeType == Interop.Sys.NodeType.DT_LNK
-                || directoryEntry.InodeType == Interop.Sys.NodeType.DT_UNKNOWN)
-                && Interop.Sys.Stat(entry.FullPath, out Interop.Sys.FileStatus targetStatus) >= 0)
+            // symlink or a directory.
+            if (isUnknown)
             {
-                // Symlink or unknown: Stat to it to see if we can resolve it to a directory.
-                isDirectory = (targetStatus.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFDIR;
+                isSymlink = entry._status.IsSymbolicLink(entry.FullPath);
+                isDirectory = entry._status.IsDirectory(entry.FullPath);
             }
+
             // Same idea as the directory check, just repeated for (and tweaked due to the
             // nature of) symlinks.
-            if (directoryEntry.InodeType == Interop.Sys.NodeType.DT_LNK)
-            {
-                isSymlink = true;
-            }
-            else if ((directoryEntry.InodeType == Interop.Sys.NodeType.DT_UNKNOWN)
-                && (Interop.Sys.LStat(entry.FullPath, out Interop.Sys.FileStatus linkTargetStatus) >= 0))
+            // Whether we had the dirent structure or not, we treat a symlink to a directory as a directory,
+            // so we need to reflect that in our isDirectory variable.
+            if (isSymlink)
```

Comment:
Otherwise, `entry._status.IsDirectory(entry.FullPath)` will be called twice for DT_UNKNOWN, right?
```suggestion
            else if (isSymlink)
```

Reply:
Good point, they are mutually exclusive. Let me add this to the issue I opened, if you don't mind.